### PR TITLE
realtek: generate CAMEO tags for DGS-1210

### PR DESCRIPTION
**Pull request by plappermaul**
Provide DGS-1210 images that will pass the U-Boot CAMEO tag check

### CODE REVIEW

File: scripts/cameo-tag.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+# rootfs partitions. If not found it complains about the image but will
+# boot. If some day we want to overwrite the original vendor partition we
+# must have the tags in place. To solve this we insert the tag two times into
+# the kernel image.
+# 
+# First location is at the end of the uImage header. This is the tricky part
+# because we have the header CRC and the CAMEO checksum that must match the
+# the whole header. Changing one of both results in a change of the other
+# So we will generate up to ~20.000 CRC and checksum combinations until we get
+# an exact match between those two.
```
[the remark this answers is not in the document]
Comment (plappermaul) [reply scripts/cameo-tag.py]: Insanity and creativity are sometimes close to each other. Comments in script have been updated.

**Comment by plappermaul** [comment]
added some clarification about the runtime of the brute force calculation

**Comment by svanheule** [comment]
Since the device will boot even if the OS partition checksum fails, can't we just add a CAMEOTAG after the FS data? That way we don't have to brute force CRCs, which would save some precious CPU cycles. The whole mutual checksum dependence for the OS partition/uImage header could still be documented in the script (or commit message), in case this ever needs to be implemented in the future.

**Comment by plappermaul** [comment]
From the forums discussion my understanding was: For successfull boot we need at least one OS+FS combination with valid checksums. I went through this synapse salad to provide these checksums for the primary partitions. This should allow us to safely expand OpenWRT to the full flash. No problems if we overwrite the CAMEO tags on the secondary partitions. If I get you right U-Boot will allow booting from primary partition set even when

kernel1 = no checksum
rootfs1 = checksum
kernel2 = no checksum
rootfs2 = no checksum.

If yes we can omit the CRC brute forcing. If not here are numbers for a "micro uImage" in a virtual machine on a i7-8550U host. Times are consistent for multiple runs

```
$ dd if=/dev/urandom of=bla.bin bs=64 count=1
1+0 data in
1+0 data out
64 Bytes copied, 0,000413525 s, 155 kB/s
$ time ~/openwrt/openwrt/scripts/cameo-tag.py bla.bin new.bin 200000
real	0m0,097s
user	0m0,070s
sys	0m0,007s
```

Remark these times are easily tripled by the fact that we copy the image (~6MB) within the calculation process.

**Comment by svanheule** [comment]
> From the forums discussion my understanding was: For successfull boot we need at least one OS+FS combination with valid checksums.

In that case we would need both checksums. From the description in the script I understood that one correct checksum out of four would sufficient to boot the system.

To optimize the CRC/checksum search, could you pre-calculate the CRC of the fixed prefix? Then you only need to feed in the guessed total byte sum for every iteration. Per the documentation of `zlib.crc32()`:
> If _value_ is present, it is used as the starting value of the checksum; otherwise, a default value of 0 is used. Passing in value allows computing a running checksum over the concatenation of several inputs.

The partial summation of the header can also be pre-calculated, since it doesn't matter if you add the CRC bytes first or last. Then you only need to add the sum of the CRC bytes to check if that matches the current guess.

**Comment by plappermaul** [comment]
I have rearranged the script. Data is changed inplace in the uImage and the checksum calculation is optimized as requested. We saved only 40% of the brute force calculation time. This makes in numbers:

```
Total script runtime: ~250 msec
Brute force part:       ~8 msec

```

**Comment by svanheule** [comment]
Sleep be damned, I kept wondering if a valid CAMEO-checksummed uImage header be calculated instead of brute-forced...

I think the answer is yes:
```python
import random

UIMAGE_HEADER = 64
UIMAGE_NAME_OFFSET = 32
CAMEO_TAG = bytearray([0x43, 0x41, 0x4d, 0x45, 0x4f, 0x54, 0x41, 0x47, 0x00, 0x00, 0x00, 0x01])
CAMEO_OFFSET = UIMAGE_NAME_OFFSET
CAMEO_CSUM_OFFSET = CAMEO_OFFSET + len(CAMEO_TAG)
CAMEO_END = CAMEO_CSUM_OFFSET + 4

def u32_toarray(value):
    return value.to_bytes(4, 'big')

fake_header = bytearray(random.randbytes(UIMAGE_HEADER))
fake_header[0:4] = u32_toarray(0x27051956) # actual uImage magic, but not so relevant here

def cameosum(buf):
    return sum(buf) & 0xffffffff # masking

def uimage_crc_prefix(buf):
    buf[4:8] = bytes([0] * 4)
    return zlib.crc32(buf)

def calc_checksums(buf):
    buf[4:8] = bytes([0xff] * 4) # set the CRC-32 to the value it will end up being
    buf[UIMAGE_NAME_OFFSET:] = bytearray([0] * 32) # clear the ih_name field

    # put the CAMEOTAG header in place
    buf[CAMEO_OFFSET:CAMEO_OFFSET + len(CAMEO_TAG)] = CAMEO_TAG

    # calculate the checksum, including the forced CRC-32, and place it in the header
    prefix_sum = cameosum(buf[:CAMEO_CSUM_OFFSET])
    buf[CAMEO_CSUM_OFFSET:CAMEO_CSUM_OFFSET + 4] = u32_toarray(prefix_sum)

    # This is where the magic happens.
    # A CRC can be checked by appending the expected value.
    # Due to the nature of the calculation, the CRC of (data + data CRC) will be zero.
    # In case of CRC-32, there is an inversion involved, so the expected result becomes 0xffffffff
    # The remaining trailing data, being zero, will have no effect on the final value.
    prefix_crc = uimage_crc_prefix(buf[:CAMEO_END])
    buf[CAMEO_END:CAMEO_END + 4] = (prefix_crc ^ 0xffffffff).to_bytes(4, 'little')
```

This results the following sort of header:
```
27 05 19 56 ff ff ff ff 49 6b 93 82 72 69 a9 9f
0f 1f 04 c9 02 09 55 cb 69 31 ab 4c ab 0a 7b a1
43 41 4d 45 4f 54 41 47 00 00 00 01 00 00 10 4d
da 11 73 3e 00 00 00 00 00 00 00 00 00 00 00 00

bytes sum: 104d
actual CRC-32: ffffffff
```

If you include this in your script, please add the following to the commit tags:
```Suggested-by: Sander Vanheule <sander@svanheule.net> # Mutual checksum algorithm```

File: scripts/cameo-tag.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+# the tag will be written somewhere to the following rootfs partition and 
+# can be found by U-Boot. The CAMEO checksum calculation must start at
+# position 0x180000 (see DLINK_KERNEL_PART_SIZE) and includes the "second"
+# half of the "split" kernel uImage. 
+
+import argparse
+import pathlib
+import socket
+import struct
+import zlib
```
Comment (svanheule) [inline scripts/cameo-tag.py:56]: Only zlib and argparse are used, the others can be dropped

File: scripts/cameo-tag.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+def write_buffer(offset, whence, buf):
+    args.uimage_file.seek(offset, whence)
+    args.uimage_file.write(buf)
+
+def toarray(value):
+    return value.to_bytes(4, 'big')
+
+def cameosum(array):
+    return sum(iter(array)) % (1<<32)
```
Comment (svanheule) [inline scripts/cameo-tag.py:74]: If this is called with something that isn't an iterable, this probably should complain anyway. So I think `sum(array) % (1<<32)` is sufficient.

File: scripts/cameo-tag.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+# tag for the second part of the kernel that resides in the
+# vendor rootfs partition. For this we will add the CAMEO tag
+# and the checksum to the end of the image.
+
+buf = read_buffer(args.rootfs_start, 0, -1)
+if len(buf) == 0:
+    raise ValueError(f"uImage smaller than " + str(args.rootfs_start) + " bytes")
+
+write_buffer(0, 2, CAMEO_TAG + toarray(cameosum(buf + CAMEO_TAG)))
```
Comment (svanheule) [inline scripts/cameo-tag.py:105]: Can you `import os` so you can use `os.SEEK_END` etc. for whence? I don't like seeing magic constants.

**Review by svanheule — COMMENTED** [review]
The title of the first commit should start with "scripts: ..."

File: scripts/cameo-tag.py (hunk 3 above)
Comment (plappermaul) [reply scripts/cameo-tag.py]: fixed

File: scripts/cameo-tag.py (hunk 2 above)
Comment (plappermaul) [reply scripts/cameo-tag.py:54]: fixed

**Comment by plappermaul** [comment]
Fixed up everthing. BTW I did not like the image name ending up with "CAMEOTAG" so I slightly changed the offsets and set the name to "OpenWrt".

**Comment by svanheule** [comment]
Thanks for the update. You managed to make the CRC calculation quite compact :slightly_smiling_face: 

After applying some cosmetic fixes, I've merged the changes into master:
e763c4c89fc5 realtek: build DGS-1210 images with CAMEO tag
f9e840b65700 scripts: add CAMEO tag generator

**Pull request closed without merging**